buildpacks/nodejs-npm-install/src/main.rs

```diff
@@ -0,0 +1,214 @@
+mod errors;
+mod layers;
+mod npm;
+
+use crate::errors::NpmInstallBuildpackError;
+use crate::layers::npm_cache::NpmCacheLayer;
+use commons::fun_run::CommandWithName;
+use commons::output::build_log::{BuildLog, Logger, SectionLogger};
+use commons::output::fmt;
+use commons::output::section_log::{log_step, log_step_stream};
+use commons::output::warn_later::WarnGuard;
+use heroku_nodejs_utils::application;
+use heroku_nodejs_utils::package_json::PackageJson;
+use heroku_nodejs_utils::package_manager::PackageManager;
+use heroku_nodejs_utils::vrs::Version;
+use libcnb::build::{BuildContext, BuildResult, BuildResultBuilder};
+use libcnb::data::build_plan::BuildPlanBuilder;
+use libcnb::data::launch::{LaunchBuilder, ProcessBuilder};
+use libcnb::data::{layer_name, process_type};
+use libcnb::detect::{DetectContext, DetectResult, DetectResultBuilder};
+use libcnb::generic::{GenericMetadata, GenericPlatform};
+use libcnb::{buildpack_main, Buildpack, Env};
+use std::io::{stdout, Stdout};
+use std::path::Path;
+use std::process::Command;
+
+pub(crate) struct NpmInstallBuildpack;
+
+impl Buildpack for NpmInstallBuildpack {
+    type Platform = GenericPlatform;
+    type Metadata = GenericMetadata;
+    type Error = NpmInstallBuildpackError;
+
+    fn detect(&self, context: DetectContext<Self>) -> libcnb::Result<DetectResult, Self::Error> {
+        context
+            .app_dir
+            .join("package.json")
+            .exists()
+            .then(|| {
+                DetectResultBuilder::pass()
+                    .build_plan(
+                        BuildPlanBuilder::new()
+                            .provides("node_modules")
+                            .requires("npm")
+                            .requires("node_modules")
+                            .requires("node")
+                            .build(),
+                    )
+                    .build()
+            })
+            .unwrap_or_else(|| DetectResultBuilder::fail().build())
```

Comment:
This silently ignores IO errors reading package.json, and instead "successfully fails" the build. ie: Instead of the build erroring out with an IO error (which would highlight the issue to the user more clearly), the buildpack will fail to detect and the user will not know why.

Reply:
yes, this was one of the questions i had asked in our weekly CNB meeting. should we be failing with an error in detect and, if so, will that error be properly surfaced to the user?

Reply:
iirc the answer in the meeting was "yes" and "yes" :-)